contracts/lens/QuoterV3.sol

```diff
@@ -0,0 +1,472 @@
+
+        return (amount, sqrtPriceX96After, initializedTicksCrossed, gasEstimate);
+    }
+
+    function quoteExactInputSingle(QuoteExactInputSingleParams memory params)
+        public
+        override
+        returns (
+            uint256 amountOut,
+            uint160 sqrtPriceX96After,
+            uint32 initializedTicksCrossed,
+            uint256 gasEstimate
+        )
+    {
+        bool zeroForOne = params.tokenIn < params.tokenOut;
+        IUniswapV3Pool pool = getPool(params.tokenIn, params.tokenOut, params.fee);
+
+        uint256 gasBefore = gasleft();
+        try
+            pool.swap(
+                address(this), // address(0) might cause issues with some tokens
+                zeroForOne,
+                params.amountIn.toInt256(),
+                params.sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1)
+                    : params.sqrtPriceLimitX96,
+                abi.encodePacked(params.tokenIn, params.fee, params.tokenOut)
+            )
+        {} catch (bytes memory reason) {
+            gasEstimate = gasBefore - gasleft();
+            return handleRevert(reason, pool, gasEstimate);
+        }
+    }
+
+    function quoteExactOutputSingle(QuoteExactOutputSingleParams memory params)
+        public
+        override
+        returns (
+            uint256 amountIn,
+            uint160 sqrtPriceX96After,
+            uint32 initializedTicksCrossed,
+            uint256 gasEstimate
+        )
+    {
+        bool zeroForOne = params.tokenIn < params.tokenOut;
+        IUniswapV3Pool pool = getPool(params.tokenIn, params.tokenOut, params.fee);
+
+        // if no price limit has been specified, cache the output amount for comparison in the swap callback
+        if (params.sqrtPriceLimitX96 == 0) amountOutCached = params.amount;
+        uint256 gasBefore = gasleft();
+        try
+            pool.swap(
+                address(this), // address(0) might cause issues with some tokens
+                zeroForOne,
+                -params.amount.toInt256(),
+                params.sqrtPriceLimitX96 == 0
+                    ? (zeroForOne ? TickMath.MIN_SQRT_RATIO + 1 : TickMath.MAX_SQRT_RATIO - 1)
+                    : params.sqrtPriceLimitX96,
+                abi.encodePacked(params.tokenOut, params.fee, params.tokenIn)
+            )
+        {} catch (bytes memory reason) {
+            gasEstimate = gasBefore - gasleft();
+            if (params.sqrtPriceLimitX96 == 0) delete amountOutCached; // clear cache
+            return handleRevert(reason, pool, gasEstimate);
+        }
+    }
+
+    /// @dev Fetch an exactin quote for a V2 pair on chain
+    /// @notice new addition
+    function quoteExactInputSingleV2(
+        uint256 amountIn,
+        address tokenIn,
+        address tokenOut
+    ) public view returns (uint256 amountOut) {
+        amountOut = this.getPairAmountOut(amountIn, tokenIn, tokenOut);
```

Comment:
why do we need `this` keyword here? was getting compiler error without it

Reply:
nvm, changed visibility of `getPairAmountIn` and `getPairAmountOut` to private and fixed it